src/Microsoft.ML.ImageAnalytics/ImagePixelExtractorTransform.cs

```diff
@@ -0,0 +1,541 @@
+                // Float: scale
+                // byte: separateChannels
+                Colors = (ColorBits)ctx.Reader.ReadByte();
+                Contracts.CheckDecode(Colors != 0);
+                Contracts.CheckDecode((Colors & ColorBits.All) == Colors);
+
+                // Count the planes.
+                int planes = (int)Colors;
+                planes = (planes & 0x05) + ((planes >> 1) & 0x05);
+                planes = (planes & 0x03) + ((planes >> 2) & 0x03);
+                Planes = (byte)planes;
+                Contracts.Assert(0 < Planes & Planes <= 4);
+
+                Convert = ctx.Reader.ReadBoolByte();
+                Offset = ctx.Reader.ReadFloat();
+                Contracts.CheckDecode(FloatUtils.IsFinite(Offset));
+                Scale = ctx.Reader.ReadFloat();
+                Contracts.CheckDecode(FloatUtils.IsFiniteNonZero(Scale));
+                Contracts.CheckDecode(Convert || Offset == 0 && Scale == 1);
+                Interleave = ctx.Reader.ReadBoolByte();
+            }
+
+            public void Save(ModelSaveContext ctx)
+            {
+                Contracts.AssertValue(ctx);
+
+#if DEBUG
+                // This code is used in deserialization - assert that it matches what we computed above.
+                int planes = (int)Colors;
+                planes = (planes & 0x05) + ((planes >> 1) & 0x05);
+                planes = (planes & 0x03) + ((planes >> 2) & 0x03);
+                Contracts.Assert(planes == Planes);
+#endif
+
+                // *** Binary format ***
+                // byte: colors
+                // byte: convert
+                // Float: offset
+                // Float: scale
+                // byte: separateChannels
+                Contracts.Assert(Colors != 0);
+                Contracts.Assert((Colors & ColorBits.All) == Colors);
+                ctx.Writer.Write((byte)Colors);
+                ctx.Writer.WriteBoolByte(Convert);
+                Contracts.Assert(FloatUtils.IsFinite(Offset));
+                ctx.Writer.Write(Offset);
+                Contracts.Assert(FloatUtils.IsFiniteNonZero(Scale));
+                Contracts.Assert(Convert || Offset == 0 && Scale == 1);
+                ctx.Writer.Write(Scale);
+                ctx.Writer.WriteBoolByte(Interleave);
+            }
+        }
+
+        internal const string Summary = "Extract color plane(s) from an image. Options include scaling, offset and conversion to floating point.";
+        internal const string UserName = "Image Pixel Extractor Transform";
+        public const string LoaderSignature = "ImagePixelExtractor";
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "IMGPXEXT",
+                //verWrittenCur: 0x00010001, // Initial
+                verWrittenCur: 0x00010002, // Swith from OpenCV to Bitmap
+                verReadableCur: 0x00010002,
+                verWeCanReadBack: 0x00010002,
```

Comment:
>0x00010002 [](start = 34, length = 10)

What would happen if you were to read a 0x00010001 model? Would it just be a *little* bit off?

Reply:
In past Loader/Resizer and Pixel extractor was tangled bundle which works together.
From what I found in my investigation during comparing System.Bitmap with old version, Pixel extractor was buggy and throw exception in case of rectangular shape (people used square shape all the time, but maybe rectangle image make sense as well) and Resizer instead of resizing image actually just pick subset of image from top left corner and count it as "Resized" image. Which is fine if you pass real width and height to transform, but if you actually want to resize them, not so great.

That's why I'm bumping versions in all 3 of them.

---
In reply to: [205264997](https://github.com/dotnet/machinelearning/pull/528#discussion_r205264997) [](ancestors = 205264997)

Reply:
Ah OK. So they were not just somewhat different, the old ones were fundamentally broken. That's nice. I guess that's a good enough reason to break backcompat.